Adding additional filtering capability

eparubrub opened the pull request:
User workflow changes: 
- Region, Industry, and Experience filter categories have been added
- Filter categories can be selected and de-selected
- Active filter labels are shown (and functional) even if filter categories are de-selected
- Shows count of all users on the right of categories
- Shows count of selected users on the right of categories
- Region is now shown as a "Pill" component to keep consistency with other filters

Code implementation changes:
- Generalizing categories of filters as "Filter"
- Simplifying logic to pull Filter data
- Redoing implementation from `FocusPicker.tsx` to `filters/FilterPicker.tsx`

Tests:
- Manually (visually) tested changes

eparubrub committed:
starting new focus picker implementation

eparubrub committed:
Merge branch 'eparubrub/fix_composition' of github.com:hawaiians/hawaiiansintech into eparubrub/add_filters

eparubrub committed:
initial circle addition to focus picker

eparubrub committed:
Merge github.com:hawaiians/hawaiiansintech into eparubrub/add_filters

eparubrub committed:
Merge branch 'eparubrub/add_filters' of github.com:hawaiians/hawaiiansintech into eparubrub/add_filters

eparubrub committed:
adding states for Picker categories
- separating Picker categories into their own component

eparubrub committed:
renaming 'FocusPicker' to 'FilterPicker'
- moving Filter components to a folder

eparubrub committed:
Merge github.com:hawaiians/hawaiiansintech into eparubrub/add_filters

eparubrub committed:
Merge branch 'eparubrub/add_filters' of github.com:hawaiians/hawaiiansintech into eparubrub/add_filters

eparubrub committed:
Showing selected filters on separate line

eparubrub committed:
adding show options after selecting Focs

eparubrub committed:
reverting selectable Focus

eparubrub committed:
testing addition to filter list

eparubrub committed:
Focus selection working
- still need to handle deselect filter when Focus is deselected

eparubrub committed:
deselect filter when Focus is deselected now working

eparubrub committed:
renaming 'Focus' to 'Filter' interface

eparubrub committed:
industries filter list appearing

eparubrub committed:
adding generic 'getFilters' function

eparubrub committed:
filtering industry and focus working

eparubrub committed:
adding filter select count

eparubrub committed:
removing getFocus and getIndustry, replacing with getFilter

eparubrub committed:
experience addition checkpoint
- adding experience filters
- debugging deselect all functionality

eparubrub committed:
simplifying experience data calling

eparubrub committed:
Region filtering working
- still need to remove region pill and highlight region present

eparubrub committed:
Merge github.com:hawaiians/hawaiiansintech into eparubrub/add_filters

eparubrub committed:
Merge branch 'eparubrub/add_filters' of github.com:hawaiians/hawaiiansintech into eparubrub/add_filters

eparubrub committed:
using Pill for region

eparubrub committed:
removing console log causing errors
- simplifying filterSelect logic on index.tsx

eparubrub committed:
cleaning comments

eparubrub committed:
initializing Focus as default

eparubrub committed:
debugging industry error in vercel deployment

eparubrub committed:
debugging stubApi error in vercel deployment

eparubrub committed:
fixing error ocurring when deselecting a filter from activeFilters when the corresponding category is closed

tellaho commented on components/filters/FilterPickerCategory.tsx:
```diff
@@ -0,0 +1,52 @@
+  return (
+    <div onClick={handleOnClick}>
+      {category}{" "}
+      <svg height="1rem" width="1rem">
```
<img width="197" alt="Screen Shot 2022-10-11 at 11 13 52 PM" src="https://user-images.githubusercontent.com/5141111/195267696-c06c8f13-2c0c-4fb7-84ed-68a74f656038.png">

Small math issue, this sizing of this SVG is slightly off. Right now it's a `1rem` container with a `.4rem` radius circle. Hence the gap between shape and its edge, at `.1rem`. You can fix this pretty easily by using percentage values for the circle's `cx`, `cy`, & `r` values ([docs](https://developer.mozilla.org/en-US/docs/Web/SVG/Element/circle)). Basically setting all three values to `50%`.

This also might be subjective, but any thoughts on bumping this down in size a bit more? I played around and liked what `0.4rem` looked like. It's likely due to how its diameter is a bit more comparable to the weight/width of the text. Any thoughts?

All together, it could look like this:

```
<svg height="0.4rem" width="0.4rem">
  <circle cx="50%" cy="50%" r="50%" />
</svg>
```

which results in:

<img width="154" alt="image" src="https://user-images.githubusercontent.com/5141111/195269190-f3ee24be-5c22-4def-a95d-27cf6fefcf74.png">

eparubrub replied:
Yea I was actually looking for a suggestion like that for the `cx`, `cy`, & `r` values. Makes sense to have the x and y values set to 50% in order to center it. 

Also, I totally agree on the sizing and I'll add this in the next commit.

tellaho commented on components/filters/FilterPickerCategory.tsx:
```diff
@@ -0,0 +1,52 @@
+  };
+  return (
+    <div onClick={handleOnClick}>
+      {category}{" "}
```
We can get away with avoiding this whitespace hack by adding margin-left to the svg below, a lá
```
svg {
  // ...
  margin-left: 0.5rem;
  // ...
}
```

eparubrub replied:
That makes sense, I'll add this in the next commit.

tellaho commented on components/filters/FilterPicker.tsx:
```diff
@@ -0,0 +1,187 @@
+  const listItemsRef = useRef<HTMLLIElement[]>([]);
+  const filterIsSelected = activeFilters.length !== 0;
+
+  useLayoutEffect(() => {
```
It looks like we removed the limit of two rows per category with the **View All** button. Thinking this `useLayoutEffect` no longer actually does anything now.

**before:**
<img width="50%" alt="image" src="https://user-images.githubusercontent.com/5141111/195274153-220dbf0d-594f-411a-8172-fe2dea9cfc8e.png">

**after:**
<img width="50%" alt="image" src="https://user-images.githubusercontent.com/5141111/195274582-fa880802-d4db-4505-a11c-5619402b1b3b.png">


The behavior was a bit awkward at times but it did limit this situation for smaller screens. Do you think it'd be worth re-adding? Happy to try advising here if you were finding it unwieldy.

eparubrub replied:
@taylorkmho yea this was interesting and something we both talked about during my initial proposals for this. I definitely had that scenario a couple times where I'm like "wait am I seeing all the filtering options? oh right there's a _View All_ option I need to press". But then again I totally see the benefit of being able to hide a lot of it for smaller screens while also showing new users we have this capability.

I tried thinking about this and had some ideas like maybe we have that same type of "limit to two rows" on initial page load, but instead of a _View All_ option, the Focus category would oscilate colors between `theme.color.brand.faded` and `theme.color.brand.base` and the 2 rowed filter options would be slightly faded as well. Then when a user clicked on it, it would open up the rest of the options and act as normal. But this implementation was a bit complex in my head and didn't want to go too deep into the implementation without talking it over.

**TLDR**: I thought about it, had some complex ideas but ultimately thought it would be better to let you advise.

tellaho replied:
If I were to throw the UX book at you, I'd lean on limiting the number of selections to something reasonable with an obvious control to increase that number. Folks have trouble keeping track of too many options.

But I'd hate for you to have to re-engineer what was built the first time. I imagine enough has changed that it's no longer 1:1 to re-add it.

eparubrub replied:
Yup, was reading about Hicks Law in **Laws of UX** on the train the other day - _The time it takes to make a decision increases with the number and complexity of choices available_. Also Miller's Law - _The average person can keep only 7 (+/- 2) items in their working memory_.

I definitely want to come back to this issue in a future PR

tellaho commented on components/MemberDirectory.tsx:
```diff
@@ -34,14 +48,36 @@ export default function MemberDirectory({ members }: MemberDirectoryProps) {
+                    <Pill
+                      active={member.regionFilter[0].active}
+                      customWidth="max-content"
+                      customFontSize="1rem"
```
Curious whatcha thinking with this new `customFontSize` prop?

<img width="694" alt="Screen Shot 2022-10-12 at 12 15 59 AM" src="https://user-images.githubusercontent.com/5141111/195275955-d7245207-335f-46e4-9253-b7ef32cc5162.png">

I worry it creates an inconsistent size between the two pill groups. It looks a bit awkward as they're so close to each other. Especially when their closest adjacent sibling (the location or job title) are otherwise the same.

eparubrub replied:
@taylorkmho I had a few things in mind with this one: 

1. I was trying to keep it a bit consistent with the sizing of the original Region text, it is slightly bigger than the filter pills on the current implementation

<img width="967" alt="Screen Shot 2022-10-12 at 8 23 51 PM" src="https://user-images.githubusercontent.com/26368729/195491862-c69f1e92-d9cb-4b02-b1c5-ea94a4a8d55f.png">

2. It feels a _tiny_ bit small when it's at the original size in comparison to the **location** right above it. In my eyes, some of the location names aren't as familiar (e.g. Bend or Lehi) with me so it feels like the **region** is more of an aside from the **location** which is visually hard to see when the **region** is much smaller, yet the **region** is the filter that groups most of the locations up. 

<img width="855" alt="Screen Shot 2022-10-12 at 8 26 47 PM" src="https://user-images.githubusercontent.com/26368729/195492243-bca564e8-1b88-4856-bd47-06b7ebff6024.png">

I don't have strong feelings about the sizing though. Maybe changing the sizing or ordering of the location might be better but either way, I'm definitely down to change it so I'll leave the decision up to you.

tellaho replied:
Any thoughts on not using a tag for the region? I understand we want to highlight selected criteria or make it obvious, which the Pill pattern comes with (with its "Selected" state). Either way, but I do agree that it does feel a bit "backwards" when it's smaller. I personally read into it as more being due to it being a tag/pill than the size.

eparubrub replied:
I don't really have a preference on whether we use a tag or not. I can experiment with a few solutions for this on the next PR.

tellaho commented on components/filters/FilterPicker.tsx:
```diff
@@ -0,0 +1,187 @@
+    setDefaultHeight(listRef.current.scrollHeight);
+  }, [width]);
+
+  function activateFilter(
```
My gut is that these look and feel like tabs, but don't behave like tabs. What do you think about adjusting the behavior to match how tabs more typically behave?

![gif](https://user-images.githubusercontent.com/5141111/195272342-b4315f8c-9029-4615-afd9-e1db15c83353.gif)

This means:
- **One (and only one) can be selected at a time.** Not more than one or zero.
- **Clicking on one deselects the others.** The only exception is when you click the active tab (which does nothing).

ataeoalii replied:
If we do allow multiple filter categories to be shown at once, it makes sense to have the most recently selected at the top instead of the bottom (so prepend the newly enabled category instead of append). This allows the user to see what they've selected especially if the screen is smaller, they'll have to scroll to the bottom.

That said, I like the suggestion of treating each category like a tab.

One minor thing, I'm definitely not a designer so take this with a grain of salt, but is there a simple way to indicate which category a selected filter belongs to? For example, if "Engineering" is selected, add a tool tip on hover stating that it belongs to the "Focus" category.

eparubrub replied:
I agree I like the tab selection, it actually also in theory keeps the user knowing what category of filter they're looking at. I'll add this to the next commit.

Also to your point Andrew, would this tab behavior make it more clear for filter selection? I think your tooltip suggestion would have been most helpful for the multiple filter categories being shown at once but since we're all in agreeance that tab behavior is better I think that solves it? Or maybe I'm not seeing some other benefits you're seeing.

eparubrub replied:
update is live here: https://hawaiiansintech-53p35qftu-hawaiians.vercel.app/

ataeoalii replied:
Yup looks good, the tool tip idea was more for the combined filter categories. I probably should've changed the order of my comments. Update looks good.

tellaho reviewed (commented):
What a sprint, brova! Full design to development.

I dropped a few comments that I think impact the UX. Let me know what you think; I'd be happy to chat about any of it. Keep it up. Mahalo for your effort.

ataeoalii commented on components/MemberDirectory.tsx:
```diff
@@ -6,6 +6,9 @@ import Pill from "./Pill";
 export interface DirectoryMember extends MemberPublic {
   focus: { active?: boolean; id: string; name: string }[];
+  industry: { active?: boolean; id: string; name: string }[];
+  experienceFilter: { active?: boolean; id: string; name: string }[];
```
One minor thing I just realized is that as this evolves over time, people's experience in tech will change year over year, so we'll probably want to change this to a date instead of a static year range (not super important to this PR tho).

eparubrub replied:
Hmmm yea this is actually an implementation design from the way we consume new user submissions, notifying @taylorkmho maybe we can discuss this in the future. I'm assuming since this site has been up, a few users already need to shift their years of experience up (assuming they didn't send an update request)

ataeoalii replied:
I added the issue here: https://github.com/hawaiiansintech/hawaiiansintech/issues/27. It may be simple enough to just keep it as is and have people manually update their profile each year.

ataeoalii commented on components/filters/FilterPicker.tsx:
```diff
@@ -0,0 +1,187 @@
+  filtersList: PickerFilter[];
+  activeFilters: PickerFilter[];
+  onFilterClick: (id?: string, filterType?: string) => any;
+  onFilterSeclect: (filterSelect?: string, enable?: boolean) => any;
```
Minor: spelling.
`onFilterSelect`

eparubrub replied:
nice catch, I'll add this correction* to the next commit.

ataeoalii commented on lib/api.ts:
```diff
@@ -164,63 +170,105 @@ export async function getMembers(): Promise<MemberPublic[]> {
+  approvedMemberIds?: string[]
+): Promise<Filter[]> {
+  const filters = await getBase({
+    name: filterType == "focus" ? "Focuses" : "Industries",
```
nit: should these filterType strings be stored in a variable somewhere?

eparubrub replied:
@ataeoalii are you suggesting something like this? 

<img width="318" alt="Screen Shot 2022-10-12 at 8 53 38 PM" src="https://user-images.githubusercontent.com/26368729/195495338-eeacceb2-34f5-4b49-b292-0e4e6b6e576c.png">

eparubrub replied:
I'm adding this to the next commit, let me know if that's not what you meant

ataeoalii reviewed (commented):
Love the added functionality, nice work. I left a few minor comments. Thanks Kamakani.

eparubrub committed:
Merge github.com:hawaiians/hawaiiansintech into eparubrub/add_filters

eparubrub committed:
Merge branch 'eparubrub/add_filters' of github.com:hawaiians/hawaiiansintech into eparubrub/add_filters

eparubrub committed:
adding a few updates based on feedback:
- filter picker category circle sizing
- spelling correction for 'onFilterSeclect'
- adding apiFilterTypes

eparubrub committed:
adding tab behavior for filter categories
- adding one more spelling correction for 'filterSelect'

ataeoalii commented on components/MemberDirectory.tsx:
```diff
@@ -14,12 +17,23 @@ interface MemberDirectoryProps {
+          .concat(mem.industry)
+          .concat(mem.experienceFilter)
+          .concat(mem.regionFilter)
+          ?.filter((foc) => foc.active).length > 0
```
Curious your thoughts here: for the selected filters, I think we should OR together within the filter category, but AND together separate categories.

For example:

If I select Region: `Hawaii` + `Florida` and Experience: `3-4 years`, I expect to see a list of people in Hawaii or Florida with 3-4 years of experience. I don't expect to see people in Hawaii with 20 years experience.

If I select Region: `Hawaii` and Experience: `3-4 years` + `5-9 years`, I expect to see all people in Hawaii with 3-9 years.

Current behavior shows people that match ANY criteria.

eparubrub replied:
Great point and something I mentioned to @taylorkmho as an idea. I thought there maybe should be a toggle between AND / OR for filter selection like you said, but then I was also thinking about the fact that it already **sorts the results by the highest amount of matched filters**, which I thought might be* enough. 

Then again I'm not too sure though as I could see some scenarios where it appears a bit confusing. The implementation wouldn't be too hard, but I'll leave that decision up to Taylor.

eparubrub replied:
Ah my bad Andrew, just now reading again, you're actually suggesting built-in AND/OR functionality which logically makes sense, I'm just having a hard time conceptualizing how we communicate this filtering logic to users via the UI/UX, or whether or not it's important to

ataeoalii replied:
Actually, that would be even better; I was just saying that the default behavior should OR within a category and AND between categories. An example where this could be helpful, say I'm a recruiter in Oregon looking for someone with 3-5 years of frontend experience seeing a list of 100 people that match any of those search criteria would be misleading when there might only be 5 people that match all of them.

tellaho replied:
Yeah after playing with it a bit, there's something about filtering by multiple categories that _feels_ like more control or communication may be ideal.

I think I'm with Andrew... I wouldn't let this be what blocks you from merging... since someone should be able to deduce how their filters relate to the results below, especially with how the tags "highlight" if they match.

ataeoalii reviewed (approved):
Approving this, feel free to ignore my 1 outstanding comment. I don't think it's critical.

tellaho commented:
I personally think you should be ready to merge unless you felt strongly about fixing some of the nits I mentioned above. We can always come back to it; it's feeling pretty dang good as it is.

eparubrub commented:
Merging with the remaining features being deferred to future PRs:
1. Limit the amount of filter options being presented when a user is first introduced to the site and provide a control to show all of the filter options. ([Refer to this comment](https://github.com/hawaiians/hawaiiansintech/pull/6#discussion_r1002770267))
2. Find a solution to handle Region filtering. Currently the Region is being displayed via Pill but has room for improvement. ([Refer to this comment](https://github.com/hawaiians/hawaiiansintech/pull/6#discussion_r1002769780))
3. Find a solution to multi-filtering controls. Currently filtering across different filter types may feel unintuitive. Some exploration with possibilities should be done with this. ([Refer to this comment](https://github.com/hawaiians/hawaiiansintech/pull/6#discussion_r994241975))
